src/circuits/safe_accumulator.rs

```diff
@@ -0,0 +1,115 @@
+use super::super::chips::safe_accumulator::{SafeAccumulatorConfig, SafeACcumulatorChip};
+use halo2_proofs::{circuit::*, halo2curves::pasta::Fp, plonk::*};
+
+#[derive(Default)]
+struct SafeAccumulatorCircuit {
+    pub values: Vec<Value<Fp>>,
+    pub accumulated_value: [Value<Fp>; 4],
+}
+
+impl Circuit<Fp> for SafeAccumulatorCircuit {
+    type Config = SafeAccumulatorConfig<4, 4>; // 4 bits for each column and 4 columns
+    type FloorPlanner = SimpleFloorPlanner;
+
+    fn without_witnesses(&self) -> Self {
+        Self::default()
+    }
+
+    fn configure(meta: &mut ConstraintSystem<Fp>) -> Self::Config {
+        let new_value = meta.advice_column();
+        let left_most_acc_inv = meta.advice_column();
+        let carry_cols = [meta.advice_column(), meta.advice_column(), meta.advice_column(), meta.advice_column()];
+        let acc_cols = [meta.advice_column(), meta.advice_column(), meta.advice_column(), meta.advice_column()];
+        let add_selector = meta.selector();
+        let overflow_selector = meta.selector();
+        let instance = meta.instance_column();
+
+        SafeACcumulatorChip::<4, 4>::configure(
+            meta,
+            new_value,
+            left_most_acc_inv,
+            carry_cols,
+            acc_cols,
+            [add_selector, overflow_selector],
+            instance,
+        )
+    }
+
+    fn synthesize(
+        &self,
+        config: Self::Config,
+        mut layouter: impl Layouter<Fp>,
+    ) -> Result<(), Error> {
+        let chip = SafeACcumulatorChip::construct(config);
+
+        let mut previous_accumulates = chip.assign(
+            layouter.namespace(|| "initial rows"),
+            0,
+            self.values[0],
+            self.accumulated_value,
+        ).unwrap();
+
+        // Actually, there is no need to multiple values for a single user.
+        // It may need multiple values who has multiple accounts in same identity
+        // so, I just keep this code for now.
+        let mut latest_accumulates: [Value<Fp>; 4];
+        for v in self.values.iter().skip(1) {
+            latest_accumulates = chip.assign(
+                layouter.namespace(|| "additional rows"),
+                0,
+                *v,
+                previous_accumulates,
+            ).unwrap();
+            previous_accumulates = latest_accumulates;
+        }
+
+        // TODO: check accumulates value with `expose_pubic` method
+        Ok(())
+    }
+}
+
+mod tests {
+    use super::SafeAccumulatorCircuit;
+    use halo2_proofs::{circuit::Value, dev::MockProver, halo2curves::{pasta::Fp, FieldExt}};
+
+    #[test]
+    fn test_none_overflow_case() {
```

Comment:
I don't understand the 2 tests that you are performing. 
What are the `accumulated_value` you define to instantiate the circuit?

Reply:
Actually, the goal of this circuit was to demonstrate matching the value and its decomposed form inside the chip. 
After ensuring they are equal, I then check for overflow. That's it.

Here `accumulated_value` is for the initial value of the accumulator. I thought that we need an example with a non-zero initialize accumulator. especially, in the case of generating proofs for many groups of users.

Now, I've modified the assign function to get decomposed values with assigned cells for result checking with `public_expose`

https://github.com/summa-dev/halo2-experiments/blob/ee1b551835460cdb0e90d1d340f9fd0a337279d4/src/chips/safe_accumulator.rs#L153-L159